proposals/state-machine-restriction-updates.md

```diff
@@ -0,0 +1,61 @@
+# State machine restriction updates
+
+## Summary
+[summary]: #summary
+
+- Allow `using (ref struct)` and `unsafe` blocks in iterators and async methods provided there is no `yield` or `await` inside the blocks.
+- Warn about `yield` inside `lock`.
+
+## Motivation
+[motivation]: #motivation
+
+It is not necessary to disallow `using` blocks with `ref struct` resources and `unsafe` blocks in async/iterator methods
+if there is no `yield` or `await` inside the blocks, because nothing from the blocks needs to be hoisted.
```

Comment:
Could this potentially apply to user-defined locals? The restrictions looks lot like what @jnm2 described here https://github.com/dotnet/csharplang/discussions/4687#discussioncomment-653724

Reply:
Yes, I'm considering that scope of the feature. Will update or add an open question. Thanks.